scripts/mod_loader/ui/widgets/root.lua

```diff
@@ -84,79 +196,76 @@ function UiRoot:event(eventloop)
 	local my = sdl.mouse.y()
 	
 	if type == sdl.events.mousewheel then
-		if self:dropdownEvent(mx, my) then
-			local done = self.currentDropDown:wheel(mx, my, eventloop:wheel())
-			table.remove(self.children)
-			if done then
-				return done
+		local pressedchild = self.pressedchild
+		local wheel = eventloop:wheel()
+
+		if pressedchild ~= nil then
+			local consumeEvent = pressedchild:wheel(mx, my, wheel)
+
+			if pressedchild.dragged and pressedchild:dragWheel(mx, my, wheel) then
+				consumeEvent = true
+			end
+
+			if consumeEvent then
+				return true
 			end
 		end
 
-		return self:wheel(mx, my, eventloop:wheel())
+		return self:wheel(mx, my, wheel)
 	end
 
 	if type == sdl.events.mousebuttondown then
 		local button = eventloop:mousebutton()
-		self:setfocus(nil)
-		local done = self:mousedown(mx, my, button)
-		if self:dropdownEvent(mx, my) then
-			done = self.currentDropDown:mousedown(mx, my, button) or done
-			table.remove(self.children)
+		local consumeEvent = false
+
+		self:releasePressedchild(mx, my, button)
+		self:setfocus(self.hoveredchild)
+
+		if button == 1 then
+			consumeEvent = self:pressHoveredchild(mx, my, button)
+		end
+
+		-- inform open dropDownUi's of mouse down event,
+		-- even if the mouse click was outside of its area,
+		-- in order to allow them to close
+		for _, dropdown in ipairs(self.dropdownUi.children) do
+			if not dropdown.containsMouse then
+				dropdown:mousedown(mx, my, button)
+			end
 		end
-		return done
+
+		self:cleanupDropdown()
+
+		return consumeEvent
```

Comment:
So if I understand the above code right, `Ui:mousedown` will never be called for mouse buttons other than 1, and it won't be possible for UI to react to `mousedown` events for buttons 2, 3, 4 and 5?

Reply:
That's a very good question!
I was actually not considering buttons 2, 4 and 5 even existing. (I don't know why I forgot about button 3 though)

My suggested solution would be:
```lua
if not consumeEvent then
	consumeEvent = self:mousedown(mx, my, button)
end
```
right before
```lua
self:cleanupDropdown()
```
Would you agree?

Reply:
Yep, sounds good